Another attempt at an astable flag

pdeffebach opened the pull request:
This attempt should be a more robust strategy than the current `@t` flag in DataFrameMacros.jl. 

I will outline it in more detail shortly.

pdeffebach commented:
More progress! Not ready for a review yet as i have not added a complete test set or begun documenting.

pdeffebach commented:
Ready for a review!

This is a big PR. In the process, I have to finally remove some deprecated functionality for grouped data frames. In particular

```
@by df :a (n1 = :a, n2 = :b)
```

used to work. In the past release we had a warning that this needed to be 

```
@by df :a $AsTable = (n1 = :a, n2 = :b)
```

and now we throw an error in the original expression. 

Here is the docstring for `@astable`

---

## `astable(args...)`

Return a `NamedTuple` from a transformation inside DataFramesMeta.jl macros.

`@astable` acts on a single block. It works through all top-level expressions
and collects all such expressions of the form `:y = x`, i.e. assignments to a
`Symbol`, which is a syntax error outside of the macro. At the end of the
expression, all assignments are collected into a `NamedTuple` to be used
with the `AsTable` destination in the DataFrames.jl transformation
mini-language.

Concretely, the expressions

```
df = DataFrame(a = 1)

@rtransform df @astable begin
    :x = 1
    y = 50
    :z = :x + y + :a
end
```

becomes the pair

```
function f(a)
    x_t = 1
    y = 50
    z_t = x_t + y + a

    (; x = x_t, z = z_t)
end

transform(df, [:a] => f => AsTable)
```

`@astable` is useful when performing intermediate calculations
yet store their results in new columns. For example, the following fails.

```
@rtransform df begin
    :new_col_1 = :x + :y
    :new_col_2 = :new_col_1 + :z
end
```

This because DataFrames.jl does not guarantee sequential evaluation of
transformations. `@astable` solves this problem

@rtransform df @astable begin
    :new_col_1 = :x + :y
    :new_col_2 = :new_col_1 + :z
end

### Examples

```
julia> df = DataFrame(a = [1, 2, 3], b = [4, 5, 6]);

julia> d = @rtransform df @astable begin
           :x = 1
           y = 5
           :z = :x + y
       end
3×4 DataFrame
 Row │ a      b      x      z
     │ Int64  Int64  Int64  Int64
─────┼────────────────────────────
   1 │     1      4      1      6
   2 │     2      5      1      6
   3 │     3      6      1      6

julia> df = DataFrame(a = [1, 1, 2, 2], b = [5, 6, 70, 80]);

julia> @by df :a @astable begin
           $(DOLLAR)"Mean of b" = mean(:b)
           $(DOLLAR)"Standard deviation of b" = std(:b)
       end
2×3 DataFrame
 Row │ a      Mean of b  Standard deviation of b
     │ Int64  Float64    Float64
─────┼───────────────────────────────────────────
   1 │     1        5.5                 0.707107
   2 │     2       75.0                 7.07107
```

---

This implementation is more complicated than that of `@t` from DataFrameMacros.jl. In DataFrameMacros.jl, the following will fail

```
df = DataFrame(a = 1)
@transform df @t begin 
    :x = 1
    b + :x 
end
```

this is because it sends `:x` to `src` in `src => fun => dest` even though it doesn't exist in the DataFrame. `@astable` does not have this problem, at the cost of a more complicated implementation. 

cc @bkamins @nalimilan

bkamins commented:
> `transform(df, [:a] => f => AsTable)`

I would think it should be `transform(df, [:a] => ByRow(f) => AsTable)`?

pdeffebach commented:
Good catch. Will update.

nalimilan commented on src/macros.jl:
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
+
+julia> df = DataFrame(a = [1, 1, 2, 2], b = [5, 6, 70, 80]);
+
+julia> @by df :a @astable begin
+           $(DOLLAR)"Mean of b" = mean(:b)
+           $(DOLLAR)"Standard deviation of b" = std(:b)
+       end
```
This example can be achieved without `@astable`, right? Maybe do `m = mean(:b); std(:b, mean=m)` to illustrate  the power of this function? Or, simpler, call `extrema(:b)` to create two columns.

Also, I wouldn't use long column names with spaces in them: better illustrate a single feature at a time.

pdeffebach replied:
great. changed.

nalimilan commented on src/parsing_astable.jl:
```diff
@@ -0,0 +1,95 @@
+function conditionally_add_symbols!(inputs_to_function, lhs_assignments, col)
```
How about making the signature more explicit? At least, `lhs_assignments::AbstractDict` would make things clearer. If you rely on the order, better even use `::OrderedCollections.OrderedDict`. Same below.

pdeffebach replied:
added.

nalimilan commented on src/parsing_astable.jl:
```diff
@@ -0,0 +1,95 @@
+function conditionally_add_symbols!(inputs_to_function, lhs_assignments, col)
+    # if it's already been assigned at top-level,
+    # don't add it to the inputs
+    if haskey(lhs_assignments, col)
+        return lhs_assignments[col]
+    else
+        return addkey!(inputs_to_function, col)
+    end
```
```suggestion
    return get!(lhs_assignments, col) do
        gensym()
    end
```

pdeffebach replied:
Ah this isn't quite right. Notice the two different dicts used, `lhs_assignments` and `inputs_to_function`.

nalimilan commented on src/parsing_astable.jl:
```diff
@@ -0,0 +1,95 @@
+                new_lhs = inputs_to_function[lhs]
+                lhs_assignments[lhs] = new_lhs
+            else
+                new_lhs = addkey!(lhs_assignments, lhs)
```
`addkey!` checks whether `lhs_assignments` contains `lhs`. If you already know that's not the case, better assign `gensym()` directly.

nalimilan commented on src/parsing.jl:
```diff
@@ -269,7 +270,13 @@ function fun_to_vec(ex::Expr;
+        return :($src => $fun => AsTable)
+    end
+
+    if no_dest # subet and with
```
```suggestion
    if no_dest # subset and with
```

nalimilan commented on test/astable_flag.jl:
```diff
@@ -0,0 +1,124 @@
+    d = @rtransform df @astable begin
+        :x = 1
+
+        nothing
```
Why return `nothing`?

pdeffebach replied:
I wanted to make sure it could handle an expression that wasn't just assignments.

nalimilan commented on src/macros.jl:
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
+Return a `NamedTuple` from a transformation inside DataFramesMeta.jl macros.
+
+`@astable` acts on a single block. It works through all top-level expressions
+and collects all such expressions of the form `:y = x`, i.e. assignments to a
```
```suggestion
and collects all such expressions of the form `:y = ...`, i.e. assignments to a
```

nalimilan commented on src/macros.jl:
````diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
+end
+```
+
+becomes the pair
````
```suggestion
become the pair
```

nalimilan commented on src/macros.jl:
````diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
+transform(df, [:a] => ByRow(f) => AsTable)
+```
+
+`@astable` is useful when performing intermediate calculations
+yet store their results in new columns. For example, the following fails.
+
+```
+@rtransform df begin
+    :new_col_1 = :x + :y
+    :new_col_2 = :new_col_1 + :z
+end
+```
+
+This because DataFrames.jl does not guarantee sequential evaluation of
+transformations. `@astable` solves this problem
````
While this is an interesting side-effect, the main goal of `AsTable` is to allow returning multiple columns from a single "function". Probably worth mentioning? For example it's useful with `extrema` to compute the minimum and the maximum at the same time.

nalimilan commented on src/macros.jl:
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
     throw(ArgumentError("@passmissing only works inside DataFramesMeta macros."))
 end
 
+"""
```
I guess you'll also add docs to the manual?

nalimilan commented on src/macros.jl:
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
+
+`@astable` acts on a single block. It works through all top-level expressions
+and collects all such expressions of the form `:y = x`, i.e. assignments to a
+`Symbol`, which is a syntax error outside of the macro. At the end of the
```
Outside of DataFramesMeta macros, right?

nalimilan reviewed (commented):
Looks pretty cool! This will require a breaking release, right?

Using `@astable` to ensure operations are run sequentially is clever. The name is a bit surprising for this, but well... I also hope the compilation overhead isn't too large.

pdeffebach commented:
> Using @astable to ensure operations are run sequentially is clever. The name is a bit surprising for this, but well... I also hope the compilation overhead isn't too large.

This macro does more than that. It allows for local variables to per persistent as well. If I were to just force sequential transform calls, I would just create a `transform` call for each `:y = f(:x)` expression.

pdeffebach commented:
This is ready for another round of reviews. I have added docs to the manual as well.

bkamins commented on docs/src/index.md:
```diff
@@ -396,11 +397,37 @@ julia> @rtransform df @passmissing x = parse(Int, :x_str)
+information. 
+
+In a single block, all assignments of the form `:y = f(:x)` 
+or `$y = f(:x)` at the top-level are generate new columns.
```
```suggestion
or `$y = f(:x)` at the top-level generate new columns.
```

bkamins replied:
maybe add what `$y` has to resolve to (I understand it has to be `Symbol`, or strings are also accepted?)

bkamins commented on src/macros.jl:
```diff
@@ -350,6 +350,120 @@ macro passmissing(args...)
 end
 
+"""
+    astable(args...)
```
```suggestion
    @astable(args...)
```

bkamins commented on src/macros.jl:
````diff
@@ -350,6 +350,120 @@ macro passmissing(args...)
+end
+```
+
+The creation of `:x_demeaned` and `:x2_demeaned` both share the variable `m`,
````
maybe add that the original data frame has to have `:x` and `:x2` columns (to make things clear).

bkamins commented on src/macros.jl:
```diff
@@ -350,6 +350,120 @@ macro passmissing(args...)
+
+Column assignment in `@astable` follows the same rules as
+column assignment more generally. Construct a new column
+from a string by escaping it with `$DOLLAR`.
```
can you add an example of this?

bkamins commented on src/macros.jl:
```diff
@@ -391,7 +392,7 @@ transform(df, [:a] => ByRow(f) => AsTable)
 First, `@astable` makes it easy to create multiple columns from a single
 transformation, which share a scope. For example, `@astable` allows
-for the following
+for the following (where `:x` and `:x_2` exist in the `DataFrame` already).
```
```suggestion
for the following (where `:x` and `:x_2` exist in the data frame already).
```

bkamins commented on docs/src/index.md:
```diff
@@ -396,11 +397,38 @@ julia> @rtransform df @passmissing x = parse(Int, :x_str)
 DataFramesMeta provides the special syntax `$` for referring to 
-columns in a data frame via a `Symbol`, string, or column position as either
-a literal or a variable. 
+columns in a data frame via a `Symbol`, string, or column position as either a literal or a variable. 
```
While we are at it given our recent discussion on Discourse, I think it is essential to mention when the `$` reference is resolved.
Also maybe add an example when macros are used within a function? I think these are cases not trivial. This can be another PR of course

pdeffebach replied:
I will do this as another PR. In summary, you can't use other macros which use `$`. I will try and sort out if I can carve out an exception.

bkamins replied:
To be clear why I stress it so much. With DataFrames.jl my answer to users is: if you learn Julia Base then you will know exactly how DataFrames.jl works. With DataFramesMeta.jl unfortunately this is not the case as it is a DSL so we need to be very precise how things work in documentation.

bkamins commented on src/macros.jl:
```diff
@@ -350,6 +350,138 @@ macro passmissing(args...)
+
+Column assignment in `@astable` follows the same rules as
+column assignment more generally. Construct a new column
+from a string by escaping it with `$DOLLAR`, which can be a
```
"from a string" here is unclear.

pdeffebach replied:
I have re-phrased. Though I think people reading this are probably familiar with column assignment so I don't want to go into it in too much detail here.

bkamins commented on src/macros.jl:
```diff
@@ -350,6 +350,138 @@ macro passmissing(args...)
+macros, `@select`, `@transform`, and their mutating and row-wise equivalents.
+
+`@astable` acts on a single block. It works through all top-level expressions
+and collects all such expressions of the form `:y = ...`, i.e. assignments to a
```
also mention the `$` form on LHS.

pdeffebach replied:
added.

bkamins commented on src/macros.jl:
```diff
@@ -350,6 +350,138 @@ macro passmissing(args...)
+
+julia> @rtransform df @astable begin
+           f_a = first(:a)
+           $(DOLLAR)new_col = :a + :b + f_a
```
where is `new_col` defined?

pdeffebach replied:
Sorry. Fixed.

nalimilan replied:
I still don't see it.

bkamins replied:
I see it. It is fixed :)

pdeffebach commented:
Thanks, I've responded to the most recent round of reviews. 

I have added more tests. I can't think of any new tests to add at the moment, they seem pretty well covered.

pdeffebach commented:
cc @jkrumbiegel, if you want to review.

nalimilan commented on src/macros.jl:
```diff
@@ -284,7 +284,7 @@ end
-    passmissing(args...)
+    @passmissing(args...)
 
 Propograte missing values inside DataFramesMeta.jl macros.
```
```suggestion
Propagate missing values inside DataFramesMeta.jl macros.
```

nalimilan commented on src/parsing_astable.jl:
```diff
@@ -43,9 +50,15 @@ function is_column_assigment(ex::Expr)
 block(ex) = isexpr(ex, :block) ? ex : :($ex;)
 
+sym_or_str_to_sym(x::Union{AbstractString, Symbol}) = Symbol(x)
+function sym_or_str_to_sym(x)
+    e = "New columns created inside @astable must be Symbols or AbstractStrings"
+    throw(ArgumentError(e))
+end
```
```suggestion
sym_or_str_to_sym(x) =
    throw(ArgumentError("New columns created inside @astable must be Symbols or AbstractStrings"))
```

nalimilan commented on test/astable_flag.jl:
```diff
@@ -120,5 +120,50 @@ end
+        :b_max = ex[2]
+    end
+
+    @test sort(d.b_min) == [5, 7]
```
This kind of test is quite fragile. Better sort the whole data frame and compare it to the reference to make sure that groups and values match.

nalimilan commented:
> This macro does more than that. It allows for local variables to per persistent as well. If I were to just force sequential transform calls, I would just create a `transform` call for each `:y = f(:x)` expression.

Yes I know. What I'm saying is that mentioning sequential operations as the main justification for it was a bit weird.

pdeffebach commented:
I don't understand the changes to th egithub review interface, but I think i've addressed lingering issues. 

I just added some checks to make sure you can't do `@passmissing` and `@astable` at the same time. I think this is intuitive in the `@byrow` case, but I want to finish #276 , which makes `@passmissing` work on column-wise transformations, before I do this.

pdeffebach commented:
Tests pass and this can be merged.

bkamins commented on test/astable_flag.jl:
```diff
@@ -0,0 +1,176 @@
+end
+
+@testset "@astable with just assignments, mutating" begin
+    # After finalizing above testset
```
@pdeffebach - this seems to be WIP?

bkamins commented on test/astable_flag.jl:
```diff
@@ -0,0 +1,178 @@
+
+    d = @rtransform df @astable begin
+        :x = 1
+
```
newline is not needed. Also I am not clear why you add `nothing` here and below? Does it change anything?

bkamins commented on test/astable_flag.jl:
```diff
@@ -0,0 +1,178 @@
+
+    x_str = "x"
+    d = @rtransform df @astable begin
+        $x_str = 1
```
this test is not good as you are assigning `1` to `:x` column, and it already was `1` so we do not see if it works correctly. Same below.

pdeffebach reviewed (commented):
Thanks for the round of reviews. I think I have addressed everything. Sorry about forgetting some of the tests.

bkamins reviewed (approved):
Looks good. Just please check why Documenter fails before merging. Thank you!

nalimilan commented on src/macros.jl:
```diff
@@ -282,11 +282,10 @@ macro byrow(args...)
     @passmissing(args...)
 
-Propograte missing values inside DataFramesMeta.jl macros.
+Propagrate missing values inside DataFramesMeta.jl macros.
```
```suggestion
Propagate missing values inside DataFramesMeta.jl macros.
```

nalimilan commented on src/macros.jl:
```diff
@@ -350,18 +349,24 @@ macro passmissing(args...)
     throw(ArgumentError("@passmissing only works inside DataFramesMeta macros."))
 end
 
+global astable_docstring_snippet = """
```
Did you mean this?
```suggestion
const astable_docstring_snippet = """
```

nalimilan commented on src/macros.jl:
```diff
@@ -350,18 +349,24 @@ macro passmissing(args...)
 end
 
+global astable_docstring_snippet = """
+    Transformations can also use the macro-flag `@astable` for creating multiple
```
```suggestion
    Transformations can also use the macro-flag [`@astable`](@ref) for creating multiple
```

nalimilan commented on src/parsing.jl:
```diff
@@ -224,6 +226,8 @@ function get_source_fun(function_expr; exprflags = deepcopy(DEFAULT_FLAGS))
         end
     end
 
+    println(MacroTools.prettify(fun))
+
```
```suggestion
```

pdeffebach commented:
Thanks for the review. @nalimilan ready to merge?

bkamins commented:
Bravo!
